File: web-app/src/lib/domain/cycle-calculations.ts
import type { CyclePhase } from "./types";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CycleSettings {
  averageCycleLengthDays: number;
  averagePeriodLengthDays: number;
  lutealPhaseLengthDays: number;
}

export interface PeriodLog {
  startDate: Date;
  endDate?: Date;
}

export interface PhaseBoundary {
  start: number;
  end: number;
}

export interface CycleStatusResult {
  currentPhase: CyclePhase;
  cycleDay: number;
  daysUntilNextPhase: number;
  predictedNextPeriod: Date;
  phaseStartDate: Date;
  phaseEndDate: Date;
}

export interface PhaseRecommendation {
  phase: CyclePhase;
  title: string;
  description: string;
  trainingFocus: string;
  intensityRecommendation: string;
  exercisesToEmphasize: string[];
  exercisesToAvoid: string[];
}

// ---------------------------------------------------------------------------
// Date helpers
// ---------------------------------------------------------------------------

function startOfDay(date: Date): Date {
  const d = new Date(date);
  d.setHours(0, 0, 0, 0);
  return d;
}

function addDays(date: Date, days: number): Date {
  const d = startOfDay(date);
  d.setDate(d.getDate() + days);
  return d;
}

function daysBetween(from: Date, to: Date): number {
  const f = startOfDay(from).getTime();
  const t = startOfDay(to).getTime();
  return Math.round((t - f) / (1000 * 60 * 60 * 24));
}

function isWithin(target: Date, start: Date, end: Date): boolean {
  const t = startOfDay(target).getTime();
  return t >= startOfDay(start).getTime() && t <= startOfDay(end).getTime();
}

// ---------------------------------------------------------------------------
// Phase Boundaries
// ---------------------------------------------------------------------------

export function getPhaseBoundaries(
  settings: CycleSettings
): Record<CyclePhase, PhaseBoundary> {
  const { averageCycleLengthDays: cycleLen, averagePeriodLengthDays: periodLen, lutealPhaseLengthDays: lutealLen } = settings;

  const ovDay = cycleLen - lutealLen;
  const ovStart = Math.max(periodLen + 2, ovDay - 2);
  const ovEnd = Math.min(ovDay + 2, cycleLen - lutealLen + 2);

  return {
    menstrual:  { start: 1,             end: periodLen },
    follicular: { start: periodLen + 1, end: ovStart - 1 },
    ovulation:  { start: ovStart,       end: ovEnd },
    luteal:     { start: ovEnd + 1,     end: cycleLen },
  };
}

// ---------------------------------------------------------------------------
// calculateCycleStatus
// ---------------------------------------------------------------------------

export function calculateCycleStatus(
  periodLogs: PeriodLog[],
  settings: CycleSettings,
  referenceDate: Date = new Date()
): CycleStatusResult | null {
  if (periodLogs.length === 0) return null;

  const ref = startOfDay(referenceDate);
  const sorted = [...periodLogs].sort(
    (a, b) => startOfDay(b.startDate).getTime() - startOfDay(a.startDate).getTime()
  );

  let cycleStartDate: Date | null = null;

  for (const period of sorted) {
    const pStart = startOfDay(period.startDate);
    const pEnd = period.endDate
      ? startOfDay(period.endDate)
      : addDays(pStart, settings.averagePeriodLengthDays - 1);

    if (isWithin(ref, pStart, pEnd)) {
      cycleStartDate = pStart;
      break;
    }
    const nextExpected = addDays(pStart, settings.averageCycleLengthDays);
    if (ref > pEnd && ref < nextExpected) {
      cycleStartDate = pStart;
      break;
    }
  }

  let cycleStart: Date;
  if (cycleStartDate) {
    cycleStart = cycleStartDate;
  } else {
    const most = sorted[0];
    let start = startOfDay(most.startDate);
    const daysSince = daysBetween(start, ref);
    const safeCycleLength = Math.max(1, settings.averageCycleLengthDays);
    const completed = Math.floor(daysSince / safeCycleLength);
    start = addDays(start, completed * safeCycleLength);
    cycleStart = start;
  }

  const cycleDay = daysBetween(cycleStart, ref) + 1;
  const boundaries = getPhaseBoundaries(settings);

  const phases: CyclePhase[] = ["menstrual", "follicular", "ovulation", "luteal"];
  let currentPhase: CyclePhase = "follicular";
  let phaseStartDay = 1;
  let phaseEndDay = settings.averageCycleLengthDays;

  for (const phase of phases) {
    const b = boundaries[phase];
    if (cycleDay >= b.start && cycleDay <= b.end) {
      currentPhase = phase;
      phaseStartDay = b.start;
      phaseEndDay = b.end;
      break;
    }
  }

  let daysUntilNext: number;
  switch (currentPhase) {
    case "menstrual":
      daysUntilNext = (boundaries.follicular?.start ?? cycleDay) - cycleDay;
      break;
    case "follicular":
      daysUntilNext = (boundaries.ovulation?.start ?? cycleDay) - cycleDay;
      break;
    case "ovulation":
      daysUntilNext = (boundaries.luteal?.start ?? cycleDay) - cycleDay;
      break;
    case "luteal":
      daysUntilNext = settings.averageCycleLengthDays - cycleDay + 1;
      break;
  }

  return {
    currentPhase,
    cycleDay,
    daysUntilNextPhase: Math.max(0, daysUntilNext),
    predictedNextPeriod: addDays(cycleStart, settings.averageCycleLengthDays),
    phaseStartDate: addDays(cycleStart, phaseStartDay - 1),
    phaseEndDate: addDays(cycleStart, phaseEndDay - 1),
  };
}

// ---------------------------------------------------------------------------
// getPhaseRecommendation
// ---------------------------------------------------------------------------

export function getPhaseRecommendation(phase: CyclePhase): PhaseRecommendation {
  switch (phase) {
    case "menstrual":
      return {
        phase: "menstrual",
        title: "Menstrual Phase",
        description: "Your period phase. Energy may be lower — you might feel more fatigued.",
        trainingFocus: "Recovery and light movement",
        intensityRecommendation: "low",
        exercisesToEmphasize: ["yoga", "walking", "light stretching"],
        exercisesToAvoid: ["heavy compound lifts", "max effort attempts"],
      };
    case "follicular":
      return {
        phase: "follicular",
        title: "Follicular Phase",
        description: "Energy and endurance begin to rise. Estrogen increases, supporting muscle growth.",
        trainingFocus: "Building strength and endurance",
        intensityRecommendation: "moderate",
        exercisesToEmphasize: ["compound movements", "strength training", "cardio"],
        exercisesToAvoid: [],
      };
    case "ovulation":
      return {
        phase: "ovulation",
        title: "Ovulation Phase",
        description: "Peak estrogen and testosterone. Often the strongest phase for performance.",
        trainingFocus: "High-intensity training and PR attempts",
        intensityRecommendation: "peak",
        exercisesToEmphasize: ["max effort attempts", "heavy compound lifts", "power-focused workouts"],
        exercisesToAvoid: [],
      };
    case "luteal":
      return {
        phase: "luteal",
        title: "Luteal Phase",
        description: "Progesterone rises, which may affect recovery and energy. Focus on maintenance.",
        trainingFocus: "Maintenance and technique refinement",
        intensityRecommendation: "moderate",
        exercisesToEmphasize: ["technique work", "volume training", "recovery-focused sessions"],
        exercisesToAvoid: ["max effort attempts", "extremely heavy loads"],
      };
  }
}

File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Cycle/CycleCalculations.swift
import Foundation

// MARK: - Cycle Phase

/// The four phases of the menstrual cycle
public enum CyclePhase: String, Codable, Sendable {
    case menstrual
    case follicular
    case ovulation
    case luteal
}

// MARK: - Types

/// Settings for cycle phase calculations
public struct CycleSettings: Codable, Sendable {
    public let averageCycleLengthDays: Int
    public let averagePeriodLengthDays: Int
    public let lutealPhaseLengthDays: Int

    public init(
        averageCycleLengthDays: Int = 28,
        averagePeriodLengthDays: Int = 5,
        lutealPhaseLengthDays: Int = 14
    ) {
        self.averageCycleLengthDays = averageCycleLengthDays
        self.averagePeriodLengthDays = averagePeriodLengthDays
        self.lutealPhaseLengthDays = lutealPhaseLengthDays
    }
}

/// A period log entry
public struct PeriodLog: Codable, Sendable {
    public let startDate: Date
    public let endDate: Date?

    public init(startDate: Date, endDate: Date? = nil) {
        self.startDate = startDate
        self.endDate = endDate
    }
}

/// Boundary days within a cycle (1-indexed)
public struct PhaseBoundary: Sendable {
    public let start: Int
    public let end: Int

    public init(start: Int, end: Int) {
        self.start = start
        self.end = end
    }
}

/// Result of cycle status calculation
public struct CycleStatusResult: Sendable {
    public let currentPhase: CyclePhase
    public let cycleDay: Int
    public let daysUntilNextPhase: Int
    public let predictedNextPeriod: Date
    public let phaseStartDate: Date
    public let phaseEndDate: Date
}

/// Training recommendation for a cycle phase
public struct PhaseRecommendation: Sendable {
    public let phase: CyclePhase
    public let title: String
    public let description: String
    public let trainingFocus: String
    public let intensityRecommendation: String
    public let exercisesToEmphasize: [String]
    public let exercisesToAvoid: [String]
}

// MARK: - Date Helpers

private let calendar = Calendar.current

private func startOfDay(_ date: Date) -> Date {
    calendar.startOfDay(for: date)
}

private func addDays(_ date: Date, _ days: Int) -> Date {
    calendar.date(byAdding: .day, value: days, to: startOfDay(date)) ?? date
}

private func daysBetween(from: Date, to: Date) -> Int {
    let fromStart = startOfDay(from)
    let toStart = startOfDay(to)
    return calendar.dateComponents([.day], from: fromStart, to: toStart).day ?? 0
}

private func isWithin(_ target: Date, start: Date, end: Date) -> Bool {
    let t = startOfDay(target)
    return t >= startOfDay(start) && t <= startOfDay(end)
}

// MARK: - Phase Boundaries

/// Calculate phase day boundaries for a given cycle settings
public func getPhaseBoundaries(settings: CycleSettings) -> [CyclePhase: PhaseBoundary] {
    getPhaseBoundaries(settings: settings, menstrualLengthOverride: nil)
}

private func getPhaseBoundaries(
    settings: CycleSettings,
    menstrualLengthOverride: Int?
) -> [CyclePhase: PhaseBoundary] {
    let cycleLen = max(1, settings.averageCycleLengthDays)
    let periodLen = min(max(1, menstrualLengthOverride ?? settings.averagePeriodLengthDays), cycleLen)
    let lutealLen = settings.lutealPhaseLengthDays

    let ovDay = cycleLen - lutealLen
    let ovStart = min(cycleLen, max(periodLen + 2, ovDay - 2))
    let ovEnd = min(cycleLen, max(ovStart, ovDay + 2))

    return [
        .menstrual:  PhaseBoundary(start: 1, end: periodLen),
        .follicular: PhaseBoundary(start: periodLen + 1, end: ovStart - 1),
        .ovulation:  PhaseBoundary(start: ovStart, end: ovEnd),
        .luteal:     PhaseBoundary(start: ovEnd + 1, end: cycleLen),
    ]
}

// MARK: - Calculate Cycle Status

/// Calculate current cycle status from period logs and settings
public func calculateCycleStatus(
    periodLogs: [PeriodLog],
    settings: CycleSettings,
    referenceDate: Date = Date()
) -> CycleStatusResult? {
    guard !periodLogs.isEmpty else { return nil }

    let ref = startOfDay(referenceDate)
    let sorted = periodLogs.sorted { startOfDay($0.startDate) > startOfDay($1.startDate) }

    var cycleStartDate: Date? = nil
    var loggedMenstrualLength: Int? = nil

    for period in sorted {
        let pStart = startOfDay(period.startDate)
        let pEnd: Date
        if let logEnd = period.endDate {
            pEnd = startOfDay(logEnd)
        } else if ref >= pStart {
            // Active period (no end date): treat as ongoing through reference date
            pEnd = ref
        } else {
            pEnd = addDays(pStart, settings.averagePeriodLengthDays - 1)
        }

        if isWithin(ref, start: pStart, end: pEnd) {
            cycleStartDate = pStart
            if period.endDate != nil {
                loggedMenstrualLength = daysBetween(from: pStart, to: pEnd) + 1
            }
            break
        }
        let nextExpected = addDays(pStart, settings.averageCycleLengthDays)
        if ref > pEnd && ref < nextExpected {
            cycleStartDate = pStart
            if period.endDate != nil {
                loggedMenstrualLength = daysBetween(from: pStart, to: pEnd) + 1
            }
            break
        }
    }

    let cycleStart: Date
    if let found = cycleStartDate {
        cycleStart = found
    } else {
        let most = sorted[0]
        var start = startOfDay(most.startDate)
        let daysSince = daysBetween(from: start, to: ref)
        let safeCycleLength = max(1, settings.averageCycleLengthDays)
        let completed = daysSince / safeCycleLength
        start = addDays(start, completed * safeCycleLength)
        cycleStart = start
    }

    let cycleDay = daysBetween(from: cycleStart, to: ref) + 1
    let boundaries = getPhaseBoundaries(settings: settings, menstrualLengthOverride: loggedMenstrualLength)

    let phases: [CyclePhase] = [.menstrual, .follicular, .ovulation, .luteal]
    var currentPhase: CyclePhase = .follicular
    var phaseStartDay = 1
    var phaseEndDay = settings.averageCycleLengthDays

    for phase in phases {
        if let b = boundaries[phase], cycleDay >= b.start && cycleDay <= b.end {
            currentPhase = phase
            phaseStartDay = b.start
            phaseEndDay = b.end
            break
        }
    }

    let daysUntilNext: Int
    switch currentPhase {
    case .menstrual:
        daysUntilNext = (boundaries[.follicular]?.start ?? cycleDay) - cycleDay
    case .follicular:
        daysUntilNext = (boundaries[.ovulation]?.start ?? cycleDay) - cycleDay
    case .ovulation:
        daysUntilNext = (boundaries[.luteal]?.start ?? cycleDay) - cycleDay
    case .luteal:
        daysUntilNext = settings.averageCycleLengthDays - cycleDay + 1
    }

    return CycleStatusResult(
        currentPhase: currentPhase,
        cycleDay: cycleDay,
        daysUntilNextPhase: max(0, daysUntilNext),
        predictedNextPeriod: addDays(cycleStart, settings.averageCycleLengthDays),
        phaseStartDate: addDays(cycleStart, phaseStartDay - 1),
        phaseEndDate: addDays(cycleStart, phaseEndDay - 1)
    )
}

// MARK: - Phase Recommendations

/// Get training recommendation for a cycle phase
public func getPhaseRecommendation(phase: CyclePhase) -> PhaseRecommendation {
    switch phase {
    case .menstrual:
        return PhaseRecommendation(
            phase: .menstrual,
            title: "Shark Week",
            description: "Your period phase. Energy may be lower — you might feel more fatigued.",
            trainingFocus: "Recovery and light movement",
            intensityRecommendation: "low",
            exercisesToEmphasize: ["yoga", "walking", "light stretching"],
            exercisesToAvoid: ["heavy compound lifts", "max effort attempts"]
        )
    case .follicular:
        return PhaseRecommendation(
            phase: .follicular,
            title: "Follicular Phase",
            description: "Energy and endurance begin to rise. Estrogen increases, supporting muscle growth.",
            trainingFocus: "Building strength and endurance",
            intensityRecommendation: "moderate",
            exercisesToEmphasize: ["compound movements", "strength training", "cardio"],
            exercisesToAvoid: []
        )
    case .ovulation:
        return PhaseRecommendation(
            phase: .ovulation,
            title: "Ovulation Phase",
            description: "Peak estrogen and testosterone. Often the strongest phase for performance.",
            trainingFocus: "High-intensity training and PR attempts",
            intensityRecommendation: "peak",
            exercisesToEmphasize: ["max effort attempts", "heavy compound lifts", "power-focused workouts"],
            exercisesToAvoid: []
        )
    case .luteal:
        return PhaseRecommendation(
            phase: .luteal,
            title: "Luteal Phase",
            description: "Progesterone rises, which may affect recovery and energy. Focus on maintenance.",
            trainingFocus: "Maintenance and technique refinement",
            intensityRecommendation: "moderate",
            exercisesToEmphasize: ["technique work", "volume training", "recovery-focused sessions"],
            exercisesToAvoid: ["max effort attempts", "extremely heavy loads"]
        )
    }
}
